colorbleed/fusion/lib.py

```diff
@@ -38,3 +45,154 @@ def update_frame_range(start, end, comp=None, set_render_range=True):
+    template_work = pipeline._format_work_template(template, session)
+
+    render_dir = os.path.join(os.path.normpath(template_work), "renders")
+    version_folder = get_next_version_folder(render_dir)
+    renders_version = os.path.join(render_dir, version_folder)
+
+    comp.Print("New renders to: %s\n" % render_dir)
+
+    with avalon.fusion.comp_lock_and_undo_chunk(comp):
+        savers = comp.GetToolList(False, "Saver").values()
+        for saver in savers:
+            filepath = saver.GetAttrs("TOOLST_Clip_Name")[1.0]
+            filename = os.path.basename(filepath)
+            new_path = os.path.join(renders_version, filename)
+            saver["Clip"] = new_path
+
+
+def switch(asset_name):
+    """Switch the current containers of the comp to the other asset (shot)
+
+    Args:
+        asset_name (str): name of the asset (shot)
+
+    Returns:
+        comp (PyObject): the comp instance
+
+    """
+
+    assert asset_name, "Function requires asset name"
+
+    host = api.registered_host()
+    assert host, "Host must be installed"
+
+    # Get current project
+    self._project = io.find_one({"type": "project",
+                                 "name": api.Session["AVALON_PROJECT"]})
+
+    # Assert asset name exists
+    # It is better to do this here then to wait till switch_shot does it
+    asset = io.find_one({"type": "asset", "name": asset_name})
+    assert asset, "Could not find '%s' in the database" % asset_name
+
+    # Use the current open comp
+    current_comp = avalon.fusion.get_current_comp()
+    assert current_comp is not None, "Could not find current comp"
+
+    containers = list(host.ls())
+    assert containers, "Nothing to update"
+
+    representations = []
+    for container in containers:
+        try:
+            representation = colorbleed.switch_item(container,
+                                                    asset_name=asset_name)
+            representations.append(representation)
+            log.debug(str(representation["_id"]) + "\n")
+        except Exception as e:
+            log.debug("Error in switching! %s\n" % e.message)
+
+    log.info("Switched %i Loaders of the %i\n" % (len(representations),
+                                                  len(containers)))
+    # Updating frame range
+    log.debug("\nUpdating frame range ..")
+    version_ids = [r["parent"] for r in representations]
+    versions = io.find({"type": "version", "_id": {"$in": version_ids}})
+    versions = list(versions)
+
+    start = min(v["data"]["startFrame"] for v in versions)
+    end = max(v["data"]["endFrame"] for v in versions)
+
+    update_frame_range(start, end, comp=current_comp)
+
+    # Build the session to switch to
+    switch_to_session = api.Session.copy()
+    switch_to_session["AVALON_ASSET"] = asset['name']
+
+    # Update session and environment
+    api.Session.update(switch_to_session)
+    os.environ.update(switch_to_session)
+
+    self._project = None
```

Comment:
Ok, so the `self._project` is only for an optimization local to this function. How about allowing to pass the `project` as argument to the methods that require them? That way it's much more explicit and also avoids the redundant complexity of storing it globally. What do you think?

Reply:
I am all up for it, no objections.